src/sources/fluent.rs

```diff
@@ -0,0 +1,1040 @@
+use super::util::{SocketListenAddr, TcpIsErrorFatal, TcpSource};
+use crate::{
+    config::{
+        log_schema, DataType, GenerateConfig, Resource, SourceConfig, SourceContext,
+        SourceDescription,
+    },
+    event::{Event, LogEvent, Value},
+    internal_events::{FluentMessageDecodeError, FluentMessageReceived},
+    tcp::TcpKeepaliveConfig,
+    tls::{MaybeTlsSettings, TlsConfig},
+};
+use bytes::{Buf, Bytes, BytesMut};
+use chrono::{serde::ts_seconds, DateTime, TimeZone, Utc};
+use flate2::read::MultiGzDecoder;
+use rmp_serde::{decode, Deserializer};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{BTreeMap, VecDeque},
+    convert::TryInto,
+    io::{self, Read},
+};
+use tokio_util::codec::Decoder;
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct FluentConfig {
+    address: SocketListenAddr,
+    tls: Option<TlsConfig>,
+    keepalive: Option<TcpKeepaliveConfig>,
+    receive_buffer_bytes: Option<usize>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<FluentConfig>("fluent")
+}
+
+impl GenerateConfig for FluentConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            address: SocketListenAddr::SocketAddr("0.0.0.0:24224".parse().unwrap()),
+            keepalive: None,
+            tls: None,
+            receive_buffer_bytes: None,
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "fluent")]
+impl SourceConfig for FluentConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        let source = FluentSource {};
+        let shutdown_secs = 30;
+        let tls = MaybeTlsSettings::from_config(&self.tls, true)?;
+        source.run(
+            self.address,
+            self.keepalive,
+            shutdown_secs,
+            tls,
+            self.receive_buffer_bytes,
+            cx.shutdown,
+            cx.out,
+        )
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "fluent"
+    }
+
+    fn resources(&self) -> Vec<Resource> {
+        vec![self.address.into()]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct FluentSource;
+
+impl TcpSource for FluentSource {
+    type Error = DecodeError;
+    type Decoder = FluentDecoder;
+
+    fn decoder(&self) -> Self::Decoder {
+        FluentDecoder::new()
+    }
+
+    fn build_event(&self, frame: FluentFrame, host: Bytes) -> Option<Event> {
+        let mut log = LogEvent::from(frame);
+        if log.get(log_schema().host_key()).is_none() {
+            log.insert(log_schema().host_key(), host);
+        }
```

Comment:
Is this better with the `entry` interface?
```suggestion
        log.entry(log_schema().host_key()).or_insert(host);
```

Reply:
👍 I always forget about that interface.